internal/cmdreporeg/command.go

```diff
@@ -0,0 +1,216 @@
+// Copyright 2022 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package cmdreporeg
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/GoogleContainerTools/kpt/internal/errors"
+	"github.com/GoogleContainerTools/kpt/internal/util/parse"
+	"github.com/GoogleContainerTools/kpt/internal/util/porch"
+	configapi "github.com/GoogleContainerTools/kpt/porch/controllers/pkg/apis/porch/v1alpha1"
+	"github.com/spf13/cobra"
+	coreapi "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/cli-runtime/pkg/genericclioptions"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+const (
+	command = "cmdreporeg"
+	longMsg = `
+kpt alpha repo reg[ister] REPOSITORY [flags]
+
+Args:
+
+REPOSITORY:
+  Address of the repository to register. Required argument.
+
+Flags:
+
+--description
+  Brief description of the package repository.
+
+--name
+  Name of the package repository. If unspecified, will use the name portion (last segment) of the repository URL.
+
+--title
+  Title of the package repository.
+
+--deployment
+  Repository is a deployment repository; packages in a deployment repository are considered deployment-ready.
+
+--repo-username
+  Username for repository authentication.
+
+--repo-password
+  Password for repository authentication.
+`
+)
+
+func newRunner(ctx context.Context, rcg *genericclioptions.ConfigFlags) *runner {
+	r := &runner{
+		ctx: ctx,
+		cfg: rcg,
+	}
+	c := &cobra.Command{
+		Use:     "reg REPOSITORY",
+		Aliases: []string{"register"},
+		Short:   "Registers a package repository with Package Orchestrator.",
+		Long:    longMsg,
+		Example: "TODO",
+		PreRunE: r.preRunE,
+		RunE:    r.runE,
+		Hidden:  porch.HidePorchCommands,
+	}
+	r.Command = c
+
+	c.Flags().StringVar(&r.title, "title", "", "Title of the package repository.")
+	c.Flags().StringVar(&r.name, "name", "", "Name of the package repository. If unspecified, will use the name portion (last segment) of the repository URL.")
+	c.Flags().StringVar(&r.description, "description", "", "Brief description of the package repository.")
+	c.Flags().BoolVar(&r.deployment, "deployment", false, "Repository is a deployment repository; packages in a deployment repository are considered deployment-ready.")
+	c.Flags().StringVar(&r.username, "repo-username", "", "Username for repository authentication.")
+	c.Flags().StringVar(&r.password, "repo-password", "", "Password for repository authentication.")
```

Comment:
qq: I don't see other flags called "username" and "password", can we use shorter flag name here? Or is there a potential risk the authentication may be ambiguous? (just curious)

Reply:
Because the Porch client interacts with kubernetes apiserver, the Porch commands pull in the generic CLI options from `cli-runtime/pkg/genericclioptions`

The `username` and `password` flags come from there and represent credentials to authenticate the client to the k8s apiserver. So i had to use different names for the credentials that will authenticate Porch to the Git repository.

This is where the flags come from:
https://github.com/kubernetes/cli-runtime/blob/master/pkg/genericclioptions/config_flags.go#L352-L357

This is where they are added to the Command:
https://github.com/GoogleContainerTools/kpt/blob/porch-cli/commands/repocmd.go#L52-L53
https://github.com/GoogleContainerTools/kpt/blob/porch-cli/commands/rpkgcmd.go#L54-L55